File: programs/ydelta/src/state/global_config.rs
//! Singleton `GlobalConfig` account that holds protocol-wide settings:
//! the protocol admin pubkey, its pending-handoff slot, and the global
//! pause switch consulted by every state-mutating instruction.

use std::mem::size_of;

use bytemuck::{Pod, Zeroable};
use shank::ShankAccount;
use solana_program::{program_error::ProgramError, pubkey::Pubkey};
use static_assertions::const_assert_eq;

use crate::require;
use crate::validation::ydelta_checker::YdeltaAccount;

/// PDA seed for the singleton [`GlobalConfig`] account.
pub const GLOBAL_CONFIG_SEED: &[u8] = b"global_config";

/// Eight-byte tag at the head of the global-config account.
pub const GLOBAL_CONFIG_DISCRIMINANT: u64 = 0x79_64_65_6C_74_61_47_63;

/// Byte size of [`GlobalConfig`].
pub const GLOBAL_CONFIG_SIZE: usize = 128;

/// Singleton account holding protocol-wide configuration: the admin
/// pubkey, its pending-transfer slot, and the protocol-wide pause flag.
#[repr(C)]
#[derive(Default, Debug, Copy, Clone, Zeroable, Pod, ShankAccount)]
pub struct GlobalConfig {
    /// Layout tag; must equal [`GLOBAL_CONFIG_DISCRIMINANT`].
    pub discriminator: u64,

    /// Current protocol admin. Signs all admin-only instructions.
    pub protocol_admin: Pubkey,

    /// Proposed next admin; takes effect via `AcceptProtocolAdmin`.
    pub pending_protocol_admin: Pubkey,

    /// Global pause switch. When non-zero, instructions that guard on it
    /// reject with `Paused`.
    pub is_paused: u8,
    _padding: [u8; 7],

    _reserved: [u64; 6],
}
const_assert_eq!(size_of::<GlobalConfig>(), GLOBAL_CONFIG_SIZE);
const_assert_eq!(size_of::<GlobalConfig>() % 8, 0);

impl YdeltaAccount for GlobalConfig {
    fn verify_discriminant(&self) -> solana_program::entrypoint::ProgramResult {
        require!(
            self.discriminator == GLOBAL_CONFIG_DISCRIMINANT,
            ProgramError::InvalidAccountData,
            "Invalid GlobalConfig discriminant: {} (expected {})",
            self.discriminator,
            GLOBAL_CONFIG_DISCRIMINANT
        )?;
        Ok(())
    }
}

impl hypertree::Get for GlobalConfig {}

impl GlobalConfig {
    /// Build a fresh config seeded with `protocol_admin`, no pending
    /// admin, and the pause switch off.
    pub fn new_empty(protocol_admin: Pubkey) -> Self {
        Self {
            discriminator: GLOBAL_CONFIG_DISCRIMINANT,
            protocol_admin,
            pending_protocol_admin: Pubkey::default(),
            is_paused: 0,
            _padding: [0; 7],
            _reserved: [0; 6],
        }
    }
}

/// Derives the singleton [`GlobalConfig`] PDA and its bump.
pub fn global_config_pda() -> (Pubkey, u8) {
    Pubkey::find_program_address(&[GLOBAL_CONFIG_SEED], &crate::ID)
}

File: ts/src/accounts/globalConfig.ts
/**
 * `GlobalConfig` — 128-byte singleton at `[b"global_config"]`. Holds the
 * protocol-wide admin and global pause flag.
 *
 * Layout (bytemuck `#[repr(C)]`):
 *   @0   8   u64    discriminator (= 0x796465_6C7461_4763)
 *   @8   32  Pubkey protocol_admin
 *   @40  32  Pubkey pending_protocol_admin
 *   @72  1   u8     is_paused
 *   @73  7   u8[7]  _padding
 *   @80  48  u64[6] _reserved
 */
import { PublicKey } from '@solana/web3.js';

import { readPubkey, readU64, readU8, view } from './_read.js';

export const GLOBAL_CONFIG_SIZE = 128;
export const GLOBAL_CONFIG_DISCRIMINANT = 0x79_64_65_6c_74_61_47_63n; // "ydeltaGc"

export interface GlobalConfig {
  protocolAdmin: PublicKey;
  pendingProtocolAdmin: PublicKey;
  isPaused: boolean;
}

export function decodeGlobalConfig(data: Uint8Array | Buffer): GlobalConfig {
  if (data.byteLength < GLOBAL_CONFIG_SIZE) {
    throw new RangeError(
      `decodeGlobalConfig: account too small (${data.byteLength} < ${GLOBAL_CONFIG_SIZE})`,
    );
  }
  const dv = view(data);
  const disc = readU64(dv, 0);
  if (disc !== GLOBAL_CONFIG_DISCRIMINANT) {
    throw new Error(
      `decodeGlobalConfig: bad discriminator 0x${disc.toString(16)} (expected 0x${GLOBAL_CONFIG_DISCRIMINANT.toString(16)})`,
    );
  }
  return {
    protocolAdmin: readPubkey(dv, 8),
    pendingProtocolAdmin: readPubkey(dv, 40),
    isPaused: readU8(dv, 72) !== 0,
  };
}
